README
NAME

    Lexical::Importer - Importer + Lexical subs/vars.

DESCRIPTION

    This is a subclass of Importer which will import all symbols as
    lexicals instead of package symbols.

IMPORTANT NOTE

    This imports symbols into the currently compiling scope which is not
    necessarily the same as the package doing the importing.

SYNOPSIS

    Say you have a module, Foo.pm:

        package Foo
    
        use base 'Exporter';
        our @EXPORT = qw/foo/;
    
        sub foo { 'not lexical' }

    You want to import foo() to use, but you also have your own foo()
    method you do not want to squash in Your::Module.pm

        # Define package versions first
        sub foo { 'not lexical' }
    
        say foo(); # prints 'not lexical';
    
        {
            use Lexical::Importer Foo => 'foo';
            say foo(); # prints 'foo'
        }
    
        say foo(); # prints 'not lexical' again;
    
        use Lexical::Importer Foo => 'foo';
        say foo(); # prints 'foo'
    
        say __PACKAGE__->foo(); # prints 'not lexical', method dispatch find package sub.
    
        # Remove lexical subs
        no Lexical::Importer;
        say foo(); # prints 'not lexical' again;

IMPORTER

    This package inherits from Importer and works exactly the same apart
    from being lexical instead of modifying the symbol table.

SEE ALSO

    Importer - The importer module this package subclasses

    Lexical::Var and Lexical::Sub - The awesome modules Zefram wrote that
    make this possible. Note: Lexical::Importer ships with a forked copy of
    these

    Lexical::Import - A similar module, but it does not support everything
    Lexical::Importer does.

LEXICAL-VAR FORK

    The Lexical::Importer module is bundled with a fork of the Lexical::Var
    XS code. This fork is necessary due to Lexical::Var being broken on
    newer perls. The author of the original package is not accepting third
    party patches, and has not yet fixed the issues himself. Once a version
    of Lexical::Var ships with a fix for newer perls this fork will likely
    be removed.

 AUTHOR

    Andrew Main (Zefram) <zefram@fysh.org>

 COPYRIGHT

    Copyright (C) 2009, 2010, 2011, 2012, 2013 Andrew Main (Zefram)
    <zefram@fysh.org>

 LICENSE

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SOURCE

    The source code repository for Lexical-Importer can be found at
    http://github.com/FIXME/.

MAINTAINERS

    Chad Granum <exodist@cpan.org>

AUTHORS

    Chad Granum <exodist@cpan.org>

COPYRIGHT

    Copyright 2016 Chad Granum <exodist@cpan.org>.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://dev.perl.org/licenses/



Changes
{{$NEXT}}

0.000005  2016-07-12 09:41:56-07:00 America/Los_Angeles

    - Fix meta-data

0.000004  2016-07-11 21:16:07-07:00 America/Los_Angeles

    - Minimum perl version 5.12 (for non-& prefixed lex subs)
    - Fix for perls older than 5.18

0.000003  2016-07-09 20:27:56-07:00 America/Los_Angeles

    - Add missing prereq

0.000002  2016-07-09 16:34:19-07:00 America/Los_Angeles

    - Documentation Fix

0.000001  2016-07-09 16:21:05-07:00 America/Los_Angeles

    - Initial Release


t/from_lex_vars/setup_c_6.pm
package t::from_lex_vars::setup_c_6;
"Lexical::Var"->import('&t6' => sub{123});
sub import { }
1;


t/from_lex_vars/setup_c_7.pm
package t::from_lex_vars::setup_c_7;
sub import { "Lexical::Var"->import('&t7' => sub{123}); }
1;


t/from_lex_vars/setup_s_6.pm
package t::from_lex_vars::setup_s_6;
"Lexical::Var"->import('$t6' => \123);
sub import { }
1;


lib/Lexical/Importer.pm
package Lexical::Importer;
use 5.012;
use warnings;
use strict;

use Lexical::SealRequireHints 0.006;
use Importer 0.013;
use parent 'Importer';

our $VERSION = "0.000004";

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

our %IMPORTED;

sub import {
    my $class = shift;
    my $from = shift;
    my $into = caller;
    $class->SUPER::import_into($from, $into, \&set_symbol, @_);
}

sub set_symbol {
    my ($name, $ref, %info) = @_;
    push @{$IMPORTED{$info{into}}} => $name if $info{sig} eq '&';
    __PACKAGE__->_import_lex_var("$info{sig}$name" => $ref);
}

sub do_unimport {
    my $self = shift;

    my $from = $self->from;
    my $imported = $IMPORTED{$from} or $self->croak("'$from' does not have any lexical imports to remove");

    my %allowed = map { $_ => 1 } @$imported;

    my @args = @_ ? @_ : @$imported;

    $self->_unimport_lex_sub($_) for @args;
}


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Lexical::Importer - Importer + Lexical subs/vars.

=head1 DESCRIPTION

This is a subclass of L<Importer> which will import all symbols as lexicals
instead of package symbols.

=head1 IMPORTANT NOTE

This imports symbols into the currently compiling scope which is not
necessarily the same as the package doing the importing.

=head1 SYNOPSIS

Say you have a module, C<Foo.pm>:

    package Foo

    use base 'Exporter';
    our @EXPORT = qw/foo/;

    sub foo { 'foo' }

You want to import C<foo()> to use, but you also have your own C<foo()> method
you do not want to squash in C<Your::Module.pm>

    # Define package versions first
    sub foo { 'not lexical' }

    say foo(); # prints 'not lexical';

    {
        use Lexical::Importer Foo => 'foo';
        say foo(); # prints 'foo'
    }

    say foo(); # prints 'not lexical' again;

    use Lexical::Importer Foo => 'foo';
    say foo(); # prints 'foo'

    say __PACKAGE__->foo(); # prints 'not lexical', method dispatch find package sub.

    # Remove lexical subs
    no Lexical::Importer;
    say foo(); # prints 'not lexical' again;

=head1 IMPORTER

This package inherits from L<Importer> and works exactly the same apart from
being lexical instead of modifying the symbol table.

=head1 SEE ALSO

L<Importer> - The importer module this package subclasses

L<Lexical::Var> and L<Lexical::Sub> - The awesome modules Zefram wrote that
make this possible. I<Note: Lexical::Importer ships with a forked copy of these>

L<Lexical::Import> - A similar module, but it does not support everything
L<Lexical::Importer> does.

=head1 LEXICAL-VAR FORK

The L<Lexical::Importer> module is bundled with a fork of the L<Lexical::Var> XS
code. This fork is necessary due to L<Lexical::Var> being broken on newer
perls. The author of the original package is not accepting third party patches,
and has not yet fixed the issues himself. Once a version of L<Lexical::Var>
ships with a fix for newer perls this fork will likely be removed.

=head2 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head2 COPYRIGHT

Copyright (C) 2009, 2010, 2011, 2012, 2013
Andrew Main (Zefram) <zefram@fysh.org>

=head2 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 SOURCE

The source code repository for Lexical-Importer can be found at
F<https://github.com/exodist/Lexical-Importer>.

=head1 MAINTAINERS

=over 4

=item Chad Granum E<lt>exodist@cpan.orgE<gt>

=back

=head1 AUTHORS

=over 4

=item Chad Granum E<lt>exodist@cpan.orgE<gt>

=back

=head1 COPYRIGHT

Copyright 2016 Chad Granum E<lt>exodist@cpan.orgE<gt>.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

See F<http://dev.perl.org/licenses/>

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.006.
use strict;
use warnings;

use 5.012000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Importer + Lexical subs/vars.",
  "AUTHOR" => "Chad Granum <exodist\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Lexical-Importer",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.012000",
  "NAME" => "Lexical::Importer",
  "PREREQ_PM" => {
    "Importer" => "0.013",
    "Lexical::SealRequireHints" => "0.006",
    "XSLoader" => 0,
    "parent" => 0
  },
  "TEST_REQUIRES" => {
    "Test2::Suite" => "0.000050"
  },
  "VERSION" => "0.000004",
  "test" => {
    "TESTS" => "t/*.t t/from_lex_vars/*.t"
  }
);


my %FallbackPrereqs = (
  "Importer" => "0.013",
  "Lexical::SealRequireHints" => "0.006",
  "Test2::Suite" => "0.000050",
  "XSLoader" => 0,
  "parent" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
